syntax = "proto3";

message JSON_value {
    oneof value {
        JSON_object obj = 1;
        JSON_array arr = 2;
        JSON_number num = 3;
        JSON_null null = 4;
        string str = 5;
    }
}

message JSON_object {
    repeated JSON_key_value_pair kv_pair = 1;
}

message JSON_key_value_pair {
    string key = 1;
    JSON_value value = 2;
}

message JSON_array {
    repeated JSON_array_item items = 1;
}

message JSON_array_item {
    JSON_value value = 1;
}

message JSON_number {
    oneof value {
        int64 long = 1;
        double double = 2;
    }
}

message JSON_null {}

message JSON_inp {
    JSON_object obj = 1;
    JSON_key_value_pair kv_pair = 2;
}

message JSON_opt {
    oneof value {
        JSON_inp inp = 1;
        JSON_object obj = 2;
    }
}